src/cloudevents/core/bindings/http.py

```diff
@@ -0,0 +1,297 @@
+        [Dict[str, Any], Optional[Union[Dict[str, Any], str, bytes]]], BaseCloudEvent
+    ],
+) -> BaseCloudEvent:
+    """
+    Parse an HTTP structured content mode message to a CloudEvent.
+
+    Deserializes the CloudEvent from the HTTP body using the specified format.
+    Any ce-prefixed headers are ignored as the body contains all event metadata.
+
+    Example:
+        >>> from cloudevents.core.v1.event import CloudEvent
+        >>> from cloudevents.core.formats.json import JSONFormat
+        >>>
+        >>> message = HTTPMessage(
+        ...     headers={"content-type": "application/cloudevents+json"},
+        ...     body=b'{"type": "com.example.test", "source": "/test", ...}'
+        ... )
+        >>> event = from_structured(message, JSONFormat(), CloudEvent)
+
+    :param message: HTTPMessage to parse
+    :param event_format: Format implementation for deserialization
+    :param event_factory: Factory function to create CloudEvent instances
+    :return: CloudEvent instance
+    """
+    return event_format.read(event_factory, message.body)
+
+
+def from_http(
+    message: HTTPMessage,
+    event_format: Format,
+    event_factory: Callable[
+        [Dict[str, Any], Optional[Union[Dict[str, Any], str, bytes]]], BaseCloudEvent
+    ],
+) -> BaseCloudEvent:
+    """
+    Parse an HTTP message to a CloudEvent with automatic mode detection.
+
+    Automatically detects whether the message uses binary or structured content mode:
+    - If any ce- prefixed headers are present → binary mode
+    - Otherwise → structured mode
+
+    This function provides a convenient way to handle both content modes without
+    requiring the caller to determine the mode beforehand.
+
+    Example:
+        >>> from cloudevents.core.v1.event import CloudEvent
+        >>> from cloudevents.core.formats.json import JSONFormat
+        >>>
+        >>> # Works with binary mode
+        >>> binary_msg = HTTPMessage(
+        ...     headers={"ce-type": "com.example.test", ...},
+        ...     body=b'...'
+        ... )
+        >>> event1 = from_http(binary_msg, JSONFormat(), CloudEvent)
+        >>>
+        >>> # Also works with structured mode
+        >>> structured_msg = HTTPMessage(
+        ...     headers={"content-type": "application/cloudevents+json"},
+        ...     body=b'{"type": "com.example.test", ...}'
+        ... )
+        >>> event2 = from_http(structured_msg, JSONFormat(), CloudEvent)
+
+    :param message: HTTPMessage to parse
+    :param event_format: Format implementation for deserialization
+    :param event_factory: Factory function to create CloudEvent instances
+    :return: CloudEvent instance
+    """
+    normalized_headers = _normalize_headers(message.headers)
+
+    for header_name in normalized_headers.keys():
+        if header_name.startswith(CE_PREFIX):
+            return from_binary(message, event_format, event_factory)
+
+    return from_structured(message, event_format, event_factory)
```

Comment:
same goes here. may be better to iterate over headers and do `lower` calls on the fly